src/Chapter3.hs

```diff
@@ -1006,6 +1119,132 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data KnightAction = AttackAction | PotionAction Health | SpellAction Defense deriving (Show, Eq)
+data MonsterAction = HitAction | RunAction deriving (Show, Eq)
+
+newtype Health = Health { unHealth :: Int } deriving (Show, Eq)
+newtype Attack = Attack { unAttack :: Int } deriving (Show, Eq)
+newtype Defense = Defense { unDefense :: Int } deriving (Show, Eq)
+
+instance Append Health where
+  append (Health h1) (Health h2) = Health (h1 + h2)
+
+instance Append Defense where
+  append (Defense d1) (Defense d2) = Defense (d1 + d2)
+
+newtype IncreaseHealth = IncreaseHealth Health
+newtype IncreaseDefense = IncreaseDefense Defense
+
+data Target = Self | Opponent
+
+data ActionResult a = Alive a | NotAlive deriving (Show, Eq)
+
+class Action a where
+  target :: a -> Target
+
+  buff :: Fighter f => f a -> a -> ActionResult (f a)
+  buff f _ = Alive f
+
+instance Action KnightAction where
+  target AttackAction = Opponent
+  target _ = Self
+
+  buff f (PotionAction h) = Alive (increaseHealth f (IncreaseHealth h))
+  buff f (SpellAction d) = Alive (increaseDefense f (IncreaseDefense d))
+  buff f _ = Alive f
+
+instance Action MonsterAction where
+  target HitAction = Opponent
+  target RunAction = Self
+
+  buff _ RunAction = NotAlive
+  buff f _ = Alive f
+
+data Monster a = Monster
+    { monsterHealth :: Health
+    , monsterAttack :: Attack
+    , monsterActions :: [a]
+    } deriving (Show, Eq)
+
+data Knight a = Knight
+    { knightHealth :: Health
+    , knightAttack :: Attack
+    , knightDefense :: Defense
+    , knightActions :: [a]
+    } deriving (Show, Eq)
+
+class Fighter f where
+  -- ehm, not sure how I did it - just tried different approaches
+  -- since actions :: (Action a) => f -> [a] won't let me create instaces
+  -- I believe the problem is the same https://stackoverflow.com/questions/44243367/rigid-type-variable-in-haskell
+  -- at this point I can take it as a belief since can't explain it
+  -- but now I think I can have as many fighters as I probably can
+  -- the only problem is that I can't restrict e.g. that Knight can only have KnightAction actions
+  actions :: (Action a) => f a -> [a]
```

Comment:
That's an interesting problem, and your solution is rather original :raised_hand_with_fingers_splayed: 
There are ways in solving this particular problem in a way you want (e.g. by restricting the type of action for different types), but they involve various language extensions.

Different people solve this problem in different ways and with different designs for this problem. One solution is to implement a `fight` method inside the `Fighter` typeclass with the `Fighter` constraint on a separate type

```haskell
class Fighter a where
    fighter :: Fighter b => a -> b -> (a, b)
```

Another way is to have actions internally and change the value based on the current action in the data type.

```haskell
class Fighter a where
    act :: a -> a
```

Maybe you can even come up with different designs for this problem without reaching to advanced features :slightly_smiling_face: It's always a good exercise to see how plain features can solve problems.